src/object_cmd.cpp

```diff
@@ -121,6 +121,7 @@ void BuildObject(ObjectType type, TileIndex tile, CompanyID owner, Town *town, u
 
 	for (TileIndex t : ta) {
 		WaterClass wc = (IsWaterTile(t) ? GetWaterClass(t) : WATER_CLASS_INVALID);
+		if (wc != WATER_CLASS_INVALID) ClearNeighbourNonFloodingStates(t);
```

Comment:
This doesn't seem to be necessary from my testings. If the object is later removed by the player, `MakeWater` resets the flooding state.